arxiv/vault/core.py

```diff
@@ -0,0 +1,217 @@
+            The time that the secret was issued.
+        lease_id : str
+            Unique ID for the lease; can be used to renew the lease for the
+            secret.
+        lease_duration : int
+            Duration of the lease in seconds, starting at ``issued``.
+        renewable : bool
+            Whether or not the lease can be renewed.
+
+        """
+        self.value = value
+        self.issued = issued
+        self.lease_id = lease_id
+        self.lease_duration = lease_duration
+        self.renewable = renewable
+
+    @property
+    def expires(self) -> datetime:
+        """Get the datetime that the lease will expire."""
+        return self.issued + timedelta(seconds=self.lease_duration)
+
+    def is_expired(self, as_of: Optional[datetime] = None) -> bool:
+        """Check whether the token is expired (as of ``as_of``)."""
+        if as_of is None:
+            as_of = datetime.now(UTC)
+        return as_of >= self.expires
+
+
+class Token(Secret):
+    """An auth token secret."""
+
+
+class Vault:
+    """A simple Vault API client."""
+
+    kubernetes_mountpoint = 'kubernetes'
+
+    def __init__(self, host: str, port: str, scheme: str = 'https',
+                 verify: Union[bool, str] = True) -> None:
+        """
+        Configure a connection to Vault.
+
+        Parameters
+        ----------
+        host : str
+            Vault host name.
+        port : str
+            Vault API port number.
+        scheme : str
+            Default is `https`.
+        verify : bool or str
+            Passed to client constructor (see :class:`hvac.v1.Client`). If a
+            bool, toggles SSL certificate verification. If a str, should be a
+            path to a certificate bundle used to verify the server certificate.
+
+        """
+        self._client = Client(url=f'{scheme}://{host}:{port}',
+                              adapter=HostnameLiberalAdapter,
+                              verify=verify)
+
+    @property
+    def client(self) -> hvac.v1.Client:
+        """Get the current HVAC Vault client."""
+        if self._client is None:
+            raise RuntimeError('No client; must authenticate')
+        return self._client
+
+    @property
+    def authenticated(self) -> bool:
+        """Indicate whether or not we are authenticated with Vault."""
+        return bool(self._client.is_authenticated())
+
+    def authenticate(self, kube_token: str, role: str) -> None:
+        """
+        Authenticate using a Kubernetes token, and obtain a Vault token.
+
+        Parameters
+        ----------
+        kube_token : str
+            Kubernetes service account token. Used to authenticate against the
+            Vault cluster using the `Kubernetes Auth Method
+            <https://www.vaultproject.io/docs/auth/kubernetes.html>`_.
+        role : str
+            Name of the Vault role against which to authenticate.
+
+        """
+        self._client.auth_kubernetes(role, kube_token,
```

Comment:
Will systems like Kinesis or lambda have a kube_token?

Reply:
Any Kinesis consumer or producer that we deploy on Kubernetes will have a service account token (``kube_token``).

If we were to use Lambda for something requiring secrets, we would need to enable an additional auth method. Here are the supported methods: https://www.vaultproject.io/docs/auth/